### PR TITLE
Open a volume

### PR DESCRIPTION
**Pull request by 0x00b1**

### CODE REVIEW

File: cellprofiler/modules/loadimages.py [hunk 1]
```diff
@@ -3183,16 +3183,20 @@ class LoadImagesImageProvider(LoadImagesImageProviderBase):
     """
 
     def __init__(self, name, pathname, filename, rescale=True,
-                 series=None, index=None, channel=None):
+                 series=None, index=None, channel=None, volume=False):
```
Comment (mcquin) [inline cellprofiler/modules/loadimages.py]: `..., volume=False, spacing=[1.0, 1.0, 1.0])`?

File: cellprofiler/modules/loadimages.py [hunk 2]
```diff
@@ -3257,20 +3261,35 @@ def provide_image(self, image_set):
...
 
+    def __provide_volume(self):
+        import skimage.io
+
+        data = skimage.io.imread(url2pathname(self.get_url()))
+
+        data = skimage.img_as_float(data)
+
+        return cpimage.Image(
+            image=data,
```
Comment (mcquin) [inline cellprofiler/modules/loadimages.py]: try: `data[:, :, :, self.channel]`

File: cellprofiler/modules/images.py [hunk 3]
```diff
@@ -172,6 +172,12 @@ def create_settings(self):
             that do not pass the current filter.
             """)
 
+        self.volumetric = cps.Binary(
+            "Load as volume(s)",
+            False,
+            doc="""Select "Yes" to load multi-page TIFs as volumes. Select "No" to load as individual frames."""
+        )
```
Comment (0x00b1) [inline cellprofiler/modules/images.py]: I think this should happen in NamesAndTypes (i.e. Select the image type).

**Comment by mcquin** [comment]
Current error happens when adding "volumetric" setting to NamesAndTypes.

We should ensure we bump the variable revision number and update upgrade_settings method.

File: cellprofiler/modules/namesandtypes.py [hunk 4]
```diff
@@ -1971,46 +2030,59 @@ def test_valid(self, pipeline, *args):
...
 
     def provide_image(self, image_set):
         image = LoadImagesImageProviderURL.provide_image(self, image_set)
-        if image.pixel_data.ndim == 2:
-            image.pixel_data = np.dstack([image.pixel_data] * 3)
+
+        if image.pixel_data.ndim == image.dimensions:
+            import skimage.color
+
+            image.pixel_data = skimage.color.gray2rgb(image.pixel_data)
```
Comment (mcquin) [inline cellprofiler/modules/namesandtypes.py]: This is the correct operation, but is a breaking change for existing users.

File: cellprofiler/modules/namesandtypes.py [hunk 5]
```diff
@@ -1971,46 +2030,59 @@ def test_valid(self, pipeline, *args):
...
     def provide_image(self, image_set):
         image = LoadImagesImageProviderURL.provide_image(self, image_set)
-        if image.pixel_data.ndim == 3:
-            image.pixel_data = \
-                np.sum(image.pixel_data, 2) / image.pixel_data.shape[2]
+
+        if image.pixel_data.ndim == image.dimensions + 1:
+            import skimage.color
+
+            image.pixel_data = skimage.color.rgb2gray(image.pixel_data)
```
Comment (mcquin) [inline cellprofiler/modules/namesandtypes.py]: This is the correct operation, but is a breaking change for existing users.

**Comment by mcquin** [comment]
@0x00b1 we're go for review

File: cellprofiler/modules/namesandtypes.py [hunk 6]
```diff
@@ -1971,46 +2037,59 @@ def test_valid(self, pipeline, *args):
...
+                                            volume=volume,
+                                            spacing=spacing)
 
     def provide_image(self, image_set):
         image = LoadImagesImageProviderURL.provide_image(self, image_set)
-        if image.pixel_data.ndim == 2:
-            image.pixel_data = np.dstack([image.pixel_data] * 3)
+
+        if image.pixel_data.ndim == image.dimensions:
+            import skimage.color
```
Comment (0x00b1) [inline cellprofiler/modules/namesandtypes.py]: Move `import` to the top of the module.

File: cellprofiler/modules/namesandtypes.py (hunk 4 above)
Comment (0x00b1) [inline cellprofiler/modules/namesandtypes.py]: I *think* it’s fine. However, I’d specify `alpha=False`.

File: cellprofiler/modules/namesandtypes.py [hunk 7]
```diff
@@ -1971,46 +2037,59 @@ def test_valid(self, pipeline, *args):
...
+                                            spacing=spacing)
 
     def provide_image(self, image_set):
         image = LoadImagesImageProviderURL.provide_image(self, image_set)
-        if image.pixel_data.ndim == 3:
-            image.pixel_data = \
-                np.sum(image.pixel_data, 2) / image.pixel_data.shape[2]
+
+        if image.pixel_data.ndim == image.dimensions + 1:
+            import skimage.color
```
Comment (0x00b1) [inline cellprofiler/modules/namesandtypes.py]: Move import to the top of the module.

File: cellprofiler/modules/namesandtypes.py (hunk 5 above)
Comment (0x00b1) [inline cellprofiler/modules/namesandtypes.py]: I think it’s fine. However, I’d specify alpha=False.

Comment (mcquin) [inline cellprofiler/modules/namesandtypes.py]: `skimage.color.rgb2gray` doesn't have an `alpha` option.

Comment (0x00b1) [inline cellprofiler/modules/namesandtypes.py]: @mcquin Oops. I misread. Ignore this.

**Pull request merged**